python/ray/train/predictor.py

```diff
@@ -124,43 +120,66 @@ def set_preprocessor(self, preprocessor: Optional[Preprocessor]) -> None:
         """Set the preprocessor to use prior to executing predictions."""
         self._preprocessor = preprocessor
 
-    def _set_cast_tensor_columns(self):
-        """Enable automatic tensor column casting.
+    @classmethod
+    @DeveloperAPI
+    def preferred_batch_format(cls) -> BatchFormat:
+        """Batch format hint for upstream producers to try yielding best block format.
+        Can be overriden by predictor classes depending on the framework type.
+
+        1. In case for DLPredictor that requires Numpy for neural network inference,
+        this method should return BatchFormat.NUMPY.
+        2. In case XGBoostPredictor that primarily works with tabular data, this
+        method can return BatchFormat.PANDAS.
 
-        If this is called on a predictor, the predictor will cast tensor columns to
-        NumPy ndarrays in the input to the preprocessors and cast tensor columns back to
-        the tensor extension type in the prediction outputs.
         """
-        self._cast_tensor_columns = True
+        return BatchFormat.PANDAS
 
     def predict(self, data: DataBatchType, **kwargs) -> DataBatchType:
         """Perform inference on a batch of data.
 
         Args:
             data: A batch of input data of type ``DataBatchType``.
             kwargs: Arguments specific to predictor implementations. These are passed
-            directly to ``_predict_pandas``.
+            directly to ``_predict_numpy`` or ``_predict_pandas``.
 
         Returns:
             DataBatchType:
                 Prediction result. The return type will be the same as the input type.
         """
-        data_df = convert_batch_type_to_pandas(data, self._cast_tensor_columns)
-
         if not hasattr(self, "_preprocessor"):
             raise NotImplementedError(
                 "Subclasses of Predictor must call Predictor.__init__(preprocessor)."
             )
-
         if self._preprocessor:
-            data_df = self._preprocessor.transform_batch(data_df)
+            data = self._preprocessor.transform_batch(data)
+        try:
+            batch_format = TYPE_TO_ENUM[type(data)]
+        except KeyError:
+            raise RuntimeError(
+                f"Invalid input data type of {type(data)}, supported "
+                f"types: {list(TYPE_TO_ENUM.keys())}"
+            )
 
-        predictions_df = self._predict_pandas(data_df, **kwargs)
-        return convert_pandas_to_batch_type(
-            predictions_df,
-            type=TYPE_TO_ENUM[type(data)],
-            cast_tensor_columns=self._cast_tensor_columns,
-        )
+        has_predict_numpy = self.__class__._predict_numpy != Predictor._predict_numpy
+
+        if batch_format == BatchFormat.PANDAS:
+            return self._predict_pandas(data, **kwargs)
```

Comment:
This will fail if `_predict_pandas` is not implemented.

Can we use the same logic as https://github.com/ray-project/ray/blob/master/python/ray/data/preprocessor.py#L186? This should cover all the cases.

Reply:
Concretely, let's say I have a Predictor with only _predict_numpy implemented. Then I call `predict` with a Pandas batch. This will fail with a `NotImplementedError`